### PR TITLE
Add FunctionNode and FunctionAdapter to the document

### PR DESCRIPTION
**Pull request by mitmul**
And fixes one trivial line

### CODE REVIEW

File: docs/source/reference/core/function_node.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FunctionNode
+------------
+
+.. autosummary::
+   :toctree: generated/
+   :nosignatures:
+
+   chainer.FunctionNode
```
Comment (unnonouno) [inline docs/source/reference/core/function_node.rst]: I feel it is better to move FunctionNode to function.rst. How do you think?

Comment (mitmul) [reply docs/source/reference/core/function_node.rst]: @unnonouno Ah, right. I moved it to function.rst and removed function_node.rst.

**Comment by unnonouno** [comment]
It contains unrelated changes i think.

**Comment by mitmul** [comment]
@unnonouno Fixed. Could you merge this?

**Comment by bonprosoft** [comment]
@mitmul I found there exists two lines which are not compliant with Sphinx autosummary in the docstring of `VariableNode`. (related #3235 )

https://github.com/chainer/chainer/blob/181186b913242efdf693ca6440f9cdb14d908aed/chainer/function_node.py#L399
https://github.com/chainer/chainer/blob/181186b913242efdf693ca6440f9cdb14d908aed/chainer/function_node.py#L444

I think they should be handled by escaping with two backslashes and one space like `w.r.t\\ `.

**Comment by mitmul** [comment]
Thanks. I think it's OK to fix them in this PR, so fixed them.

File: chainer/function_node.py [hunk 2]
```diff
@@ -238,7 +238,7 @@ def apply(self, inputs):
             # Topological ordering
             self.rank = max([x.rank for x in input_vars]) if input_vars else 0
             # Add backward edges
-            for y in ret:
+            for i, y in enumerate(ret):
```
Comment (beam2d) [inline chainer/function_node.py]: Is this change relevant?

**Comment by beam2d** [comment]
LGTM

**Pull request merged**